src/legacy/server/logging/rotate/log_rotator.ts

```diff
@@ -0,0 +1,364 @@
+    this.isRotating = true;
+
+    // get rotated files metadata
+    const foundRotatedFiles = await this._readRotatedFilesMetadata();
+
+    // delete number of rotated files exceeding the keepFiles limit setting
+    const rotatedFiles: string[] = await this._deleteFoundRotatedFilesAboveKeepFilesLimit(
+      foundRotatedFiles
+    );
+
+    // delete last file
+    await this._deleteLastRotatedFile(rotatedFiles);
+
+    // rename all files to correct index + 1
+    // and normalize
+    await this._renameRotatedFilesByOne(rotatedFiles);
+
+    // rename current log into 0
+    await this._rotateCurrentLogFile();
+
+    // send SIGHUP to reload log configuration
+    this._sendReloadLogConfigSignal();
+
+    // Reset log file size
+    this.logFileSize = 0;
+
+    // rotate process is finished
+    this.isRotating = false;
+  }
+
+  async _readRotatedFilesMetadata() {
+    const logFileBaseName = basename(this.logFilePath);
+    const logFilesFolder = dirname(this.logFilePath);
+    const foundLogFiles: string[] = await readdirAsync(logFilesFolder);
+
+    return (
+      foundLogFiles
+        .filter(file => new RegExp(`${logFileBaseName}\\.\\d`).test(file))
+        // we use .slice(-1) here in order to retrieve the last number match in the read filenames
+        // @ts-ignore here as the warning is not too important
+        .sort((a, b) => Number(a.match(/(\d+)/g).slice(-1)) - Number(b.match(/(\d+)/g).slice(-1)))
+        .map(filename => `${logFilesFolder}${sep}${filename}`)
+    );
+  }
+
+  async _deleteFoundRotatedFilesAboveKeepFilesLimit(foundRotatedFiles: string[]) {
+    if (foundRotatedFiles.length <= this.keepFiles) {
+      return foundRotatedFiles;
+    }
+
+    const finalRotatedFiles = foundRotatedFiles.slice(0, this.keepFiles);
+    const rotatedFilesToDelete = foundRotatedFiles.slice(
+      finalRotatedFiles.length,
+      foundRotatedFiles.length
+    );
+
+    await Promise.all(
+      rotatedFilesToDelete.map((rotatedFilePath: string) => unlinkAsync(rotatedFilePath))
+    );
+
+    return finalRotatedFiles;
+  }
+
+  async _deleteLastRotatedFile(rotatedFiles: string[]) {
+    if (rotatedFiles.length < this.keepFiles) {
+      return;
+    }
+
+    const lastFilePath: string = rotatedFiles.pop() as string;
+    await unlinkAsync(lastFilePath);
+  }
+
+  async _renameRotatedFilesByOne(rotatedFiles: string[]) {
+    const logFileBaseName = basename(this.logFilePath);
+    const logFilesFolder = dirname(this.logFilePath);
+
+    for (let i = rotatedFiles.length - 1; i >= 0; i--) {
+      const oldFilePath = rotatedFiles[i];
+      const newFilePath = `${logFilesFolder}${sep}${logFileBaseName}.${i + 1}`;
+      await renameAsync(oldFilePath, newFilePath);
+    }
+  }
+
+  async _rotateCurrentLogFile() {
+    const newFilePath = `${this.logFilePath}.0`;
+    await renameAsync(this.logFilePath, newFilePath);
+  }
+
+  _sendReloadLogConfigSignal() {
+    if (isMaster) {
+      // @ts-ignore
+      process.emit('SIGHUP');
```

Comment:
Why TS requires the second argument? to send a real signal? 
```ts
process.emit('SIGHUP', 'SIGHUP');
```

Reply:
just a problem with types, workaround added